[ADD] Project Template Module

patrickrwilson opened the pull request:
Adds Project Template Module

Original Discussion #500

patrickrwilson committed:
[ADD] Project Template Module

Adding new Project Template Module

[FIX] PEP8 Errors

Revert "[FIX] PEP8 Errors"

This reverts commit 90f9cd939a72ece1fbfd8f42901cf479a66ac10c.

[FIX] Fixed PEP8 Formatting

First upload tests showed some lint issues which this commit has the fixes for.

[FIX] Fixed LINT Issue

Removed unreachable code that was not needed.

[FIX] RunBot Warning

Fixed runbot warning about @class to hasclass

[UPD] Suggested Changes

Updated based on suggested changes on other submitted modules.

[UPD] Manifest and Readme

Updated Manifest and Readme based on comments on other module pr's.

[UPD] Suggested Changes

Updated code based on suggested code changes from reviewer.

[FIX] Creation Error

Fixed create project error when creating project via configuration/projects menu

[UPD] Menu and Bool Change

Removed Template Menu (Filter option still there) and moved Bool up below name. (From reviewer suggestions)

[UPD] Kanban Menu

Removed button, added link to kanban menu. Start of tests

[UPD] Remove project_milestone code

Removed the project_milestone code and will create a separate module for that purpose.

dreispt commented on project_template/models/project.py:
```diff
@@ -0,0 +1,70 @@
+            new_project.subtask_project_id = new_project.id
+        # Clear Email Alias if it was set in the Template
+        if new_project.alias_name:
+            new_project.alias_name = False
```
maybe we can set this on the copy default, at line 15.

patrickrwilson replied:
Good point, I removed lines 20-22 and added this to the copy function. Seems to work as intended. Thanks.

dreispt commented on project_template/models/project.py:
```diff
@@ -0,0 +1,70 @@
+    @api.onchange('is_template')
+    def on_change_is_template(self):
+        # Add "(TEMPLATE)" to the Name if is_template == true
+        if self.name:
```
I feel that this `if` statement is not needed.

patrickrwilson replied:
I think you are right, removing it had no impact on functionality so i've removed it. Thanks.

rafaelbn commented:
I think is better idea to restore this functionality

https://github.com/odoo/odoo/blob/8.0/addons/project/project.py#L281

Add a Boolean for templates is not bad an is simple, but imagine adding more Boolean for more similar and simple purposes? :smile: 

It is just my opinion for a more general approach.

Not blocking

tarteo reviewed (commented):
Would be nice to be able to make a template from a project (reversed create_project_from_template).

tbaden commented:
by coincidence we did something similar for v10 https://github.com/OCA/project/pull/512 😅

patrickrwilson commented:
@rafaelbn I'm not too familiar with project functionalities of older versions of Odoo but I do see the appeal of some of those features. I don't think it would be too much to improve this module with additional features. The boolean was a simple, straight forward way to manage the templates which is why I chose that route. I'm sure there are many other options out there, I just chose one that was similar to how our Dynamics 365 project solution works.

@tarteo This is a good idea for a possible improvement. Currently you can change any project to a template simply by the boolean field. To keep the project you want to create a template from, you would have to duplicate it first then change to template. I can see this function having a bit more logic as you probably would want it to 'reset' the tasks and such as if they are from a current project they would have additional data in them. I think this would require some discussion on the best routes.

@tbaden I see that pr was never merged, perhaps it wasn't fully completed?

tbaden commented:
If you go to "Configuration" -> "Projects" -> "Create Project" 
![Bildschirmfoto 2019-04-26 um 16 03 01](https://user-images.githubusercontent.com/6755125/56813184-cd1bfa00-683c-11e9-98a5-2e1c9089591e.png)

tbaden commented:
@patrickrwilson no we created it just this week and today we created the PR

patrickrwilson commented:
@tbaden nice catch, I missed creating the project that way. I'll work on the fix. Thanks

This has been fixed, I recently removed a if statement "if self.name:" in that function cause a reviewer and I thought it wasn't needed, however now I know why I originally had it in there. The error shouldn't show anymore. :)

patrickrwilson commented:
Fixed creation error and squashed commits. Thanks for your reviews. Please let me know what else might be needed.

rafaelbn commented:
It's OK. I just pointed that the functionality was in Odoo v8, more complete as @tarteo said, because this could help is very important to have back and front view in the ecosystem :smile: :

![2019-04-28_2-16-27](https://user-images.githubusercontent.com/8736623/56856491-ea93b580-695b-11e9-951f-b6f623bdc79e.png)


I understand you, and it's OK but i want to explain that the solutions in the OCA should be thunk as much generic as possible with the idea that more companies can use it

> I just chose one that was similar to how our Dynamics 365 project solution works

Usually this process of thinking which is the more generic and expandable way to do something for more customers and companies make better solutions

Do not get me wrong, I do not want to discourage you, much less. I thank you and I am very happy to have a new contributor in OCA but I wanted to give you my point of view.

After testing as is:

- I think menu "Templates" is not necessary, it take a really big space in the menu for just a little functionality. Plus you have exacly the same in the first menu changing filter.

![2019-04-28_2-38-15](https://user-images.githubusercontent.com/8736623/56856682-f08b9580-695f-11e9-9fab-2bd4401bce62.png)

- The Boolean for is template could be added in other place? Like right side of the other Boolean "Allow timesheets" or inside the Tab "Settings" inside the "Configuration" section

![2019-04-28_2-42-33](https://user-images.githubusercontent.com/8736623/56856702-7c052680-6960-11e9-8d97-8cf34ed08a36.png)

or

![2019-04-28_2-48-39](https://user-images.githubusercontent.com/8736623/56856705-81627100-6960-11e9-86fd-529362ce1e74.png)

patrickrwilson commented:
@rafaelbn great suggestions, I think these are good improvements and will make these changes shortly.

patrickrwilson commented:
@rafaelbn I removed the template menu item and it does look much cleaner. I moved the bool field up below the name which looks nicer. I didn't put it in the configuration area since, at least on my instance, it required debug mode to see that. I do think your suggested changes did give it a cleaner look. Thanks

rafaelbn commented:
Hi @patrickrwilson , thank you!

It really needed test `codecov/patch — 17.64%` if don't know how to achive this, please let us know @Yajo or @pedrobaeza will point you some info so you can learn.

By the way, maybe you don't know it, we use this module in all our instances of Odoo: https://github.com/OCA/server-ux/tree/12.0/base_technical_features . This show some needed stuff for the users that Odoo hide and only is shown in Debug mode as you said.

Regards

rafaelbn commented:
How to make test, I found this @patrickrwilson 

- https://www.odoo.com/documentation/12.0/reference/testing.html
- https://www.odoo.yenthevg.com/creating-automated-tests-odoo/

rafaelbn commented:
@tarteo I think you can review this. It's quite simple and easy to test.

> Would be nice to be able to make a template from a project (reversed create_project_from_template).

@patrickrwilson add this to Known issues

IMHO with test is Ready to merge

@dreispt what do you think? :smile:

rafaelbn reviewed (commented):
Missing test

patrickrwilson commented:
@rafaelbn I'm working on tests, might take me a bit to figure it out but thanks for the links.

@tarteo I can add a simple 'Create template from project' button but it may need further discussion. If we use the copy method, then we would need to 'clean' the tasks as you wouldn't want all the junk that gets added to them if the project wasn't new. Maybe i'm thinking too much into it or there is another way you were thinking?

tarteo reviewed (changes requested):
It works. The project_key module has a constraint on key so to make a project from a template you need to change the key to something unique. The only thing for me is this button:
![image](https://user-images.githubusercontent.com/10028499/56947182-9f7ccc80-6b2c-11e9-87c0-ef288f69a054.png)
It doesn't match Odoo's style. Maybe you can put the button here:
![image](https://user-images.githubusercontent.com/10028499/56947243-ce933e00-6b2c-11e9-951f-df9a15dbc129.png)
This way the user can't accidentally click it and looks good:

patrickrwilson commented:
@tarteo I'm fine with making the change on the button. I'm all for consistency and clean looks. I'll have that in my next commit.

I can add a check in there that if the project_key module is installed, it automatically changes the key to be used but i'm not sure what to. Maybe a random number? Otherwise it can change it to something that already exists. Ideas?

tarteo commented:
@patrickrwilson I think it is not a problem of this module, it should be changed in project_key. I think I already made the fix in 12.0 https://github.com/OCA/project/pull/426

patrickrwilson commented:
@tarteo sounds good, I do have it working with adding it in the card menu instead of the button, so that will be in the next commit. I'm currently struggling with the unit tests and will probably have quite a few more commits for that soon.

patrickrwilson commented:
I'm still stuck on the test, my last commit has the menu change, but i'm just not sure how to build out the test and could use your expertise.

tarteo reviewed (approved):
Thanks!

tarteo commented on project_template/tests/test_project_template.py:
```diff
@@ -0,0 +1,25 @@
+
+    # TEST 01: Set project to be a template
+    def test_template(self):
+        test_project.is_template = True
```
test_project cannot to be reached from here it's out of scope. You need to make an instance variable using ```self```.  In ```setUp``` do self.test_project instead.

To run unit tests locally add ```--test-enable -u project_template -d {mydb}```. Replace {mydb} with your database name. You can find more info here: https://www.odoo.com/documentation/11.0/reference/testing.html

patrickrwilson commented:
Thanks for everyone's help. I'm slowly getting better at unit tests and this is much closer. 

I have a piece of code in there that i'm not sure how to put into the unit test. If the project_milestone module is also installed (PR #501) then it will link the new tasks to the new milestones when a project is created from a template. I had to do this since deepcopy was a dangerous option I didn't feel like I should dive into when creating this. It ignores that code if the milestone module is not installed.

Does anyone know how to add some testing in unit tests for this? Suggestions?

max3903 commented:
@patrickrwilson You can create a `project_template_milestone` module with:
* your piece of code
* `depends = project_template, project_milestone`
* `auto_install = True`

patrickrwilson commented:
@max3903 that makes sense and seems to be the proper way to do it. I've removed the code and will work on another module that implements that. Thanks. I've squashed the commits.
